### PR TITLE
Add Avalanche, Benqi connectors

### PR DESCRIPTION
**Pull request by babkendev01**
- Add Avalanche common contracts
- Add connector for Benqi on Avalanche

### CODE REVIEW

File: contracts/avalanche/common/interfaces.sol [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+interface MemoryInterface {
+    function getUint(uint id) external returns (uint num);
+    function setUint(uint id, uint val) external;
+}
+
+interface InstaMapping {
+    function qiTokenMapping(address) external view returns (address);
+    function gemJoinMapping(bytes32) external view returns (address);
+}
```
Comment (thrilok209) [inline contracts/avalanche/common/interfaces.sol]:
```suggestion
```

File: contracts/avalanche/common/stores.sol [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+  /**
+   * @dev Return Wrapped AVAX address
+   */
+  address constant internal wavaxAddr = 0xB31f66AA3C1e785363F0875A1B74E27b85FD66c7;
+
+  /**
+   * @dev Return memory variable address
+   */
+  MemoryInterface constant internal instaMemory = MemoryInterface(0x8a5419CfC711B2343c17a6ABf4B2bAFaBb06957F);
```
Comment (thrilok209) [inline contracts/avalanche/common/stores.sol]:
```suggestion
  /**
   * @dev Return memory variable address
   */
  MemoryInterface constant internal instaMemory = MemoryInterface(0x3254Ce8f5b1c82431B8f21Df01918342215825C2);
```

File: contracts/avalanche/common/stores.sol [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+  /**
+   * @dev Return InstaDApp Mapping Addresses
+   */
+  InstaMapping constant internal instaMapping = InstaMapping(0xe81F70Cc7C0D46e12d70efc60607F16bbD617E88);
+
```
Comment (thrilok209) [inline contracts/avalanche/common/stores.sol]:
```suggestion
```

File: contracts/avalanche/mapping/benqi.sol [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+interface ConnectorsInterface {
+    function chief(address) external view returns (bool);
+}
+
+interface QiTokenInterface {
+    function isQiToken() external view returns (bool);
+    function underlying() external view returns (address);
+}
+
```
Comment (thrilok209) [inline contracts/avalanche/mapping/benqi.sol:16]:
```suggestion
interface MappingControllerInterface {
    function hasRole(address,address) external view returns (bool);
}

```

File: contracts/avalanche/mapping/benqi.sol [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    IndexInterface public constant instaIndex = IndexInterface(0x2971AdFa57b20E5a416aE5a708A8655A9c74f723);
+
+    address public constant avaxAddr = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+    mapping (string => TokenMap) public qiTokenMapping;
+
+    modifier isChief {
+        require(msg.sender == instaIndex.master() || connectors.chief(msg.sender), "not-an-chief");
+        _;
+    }
```
Comment (thrilok209) [inline contracts/avalanche/mapping/benqi.sol:46]:
```suggestion
    modifier isChief {
        require(msg.sender == instaIndex.master() || connectors.chief(msg.sender), "not-an-chief");
        _;
    }
    
    modifier hasRoleOrIsChief {
        require(
            msg.sender == instaIndex.master() ||
                connectors.chief(msg.sender) ||
                mappingController.hasRole(address(this), msg.sender),
            "not-an-chief/controller"
        );
        _;
    }
```

File: contracts/avalanche/mapping/benqi.sol [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        address token;
+    }
+
+    event LogQiTokenAdded(string indexed name, address indexed token, address indexed qitoken);
+    event LogQiTokenUpdated(string indexed name, address indexed token, address indexed qitoken);
+
+    ConnectorsInterface public immutable connectors;
+
+    // InstaIndex Address.
+    IndexInterface public constant instaIndex = IndexInterface(0x2971AdFa57b20E5a416aE5a708A8655A9c74f723);
```
Comment (thrilok209) [inline contracts/avalanche/mapping/benqi.sol]:
```suggestion
    // InstaConnectorsV2
    ConnectorsInterface public constant connectors = ConnectorsInterface(0x127d8cD0E2b2E0366D522DeA53A787bfE9002C14);

    // InstaIndex Address.
    IndexInterface public constant instaIndex = IndexInterface(0x6CE3e607C808b4f4C26B7F6aDAeB619e49CAbb25);
    
     // InstaMappingController Address.
     MappingControllerInterface public constant mappingController = MappingControllerInterface(0xF2113d0c99f36D7D6F6c6FAf05E0863892255999);
```

File: contracts/avalanche/mapping/benqi.sol [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            emit LogQiTokenAdded(_names[i], _tokens[i], _qitokens[i]);
+        }
+    }
+
+    function updateQitokenMapping(
+        string[] calldata _names,
+        address[] memory _tokens,
+        address[] calldata _qitokens
+    ) external {
+        require(msg.sender == instaIndex.master(), "not-master");
```
Comment (thrilok209) [inline contracts/avalanche/mapping/benqi.sol]:
```suggestion
    ) external IsChief {
```

File: contracts/avalanche/mapping/benqi.sol [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            );
+            emit LogQiTokenUpdated(_names[i], _tokens[i], _qitokens[i]);
+        }
+    }
+
+    function addQitokenMapping(
+        string[] memory _names,
+        address[] memory _tokens,
+        address[] memory _qitokens
+    ) external isChief {
```
Comment (thrilok209) [inline contracts/avalanche/mapping/benqi.sol]:
```suggestion
    ) external hasRoleOrIsChief {
```

File: contracts/avalanche/mapping/benqi.sol [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    string constant public name = "Benqi-Mapping-v1.0";
+
+    constructor(
+        address _connectors,
+        string[] memory _qitokenNames,
+        address[] memory _tokens,
+        address[] memory _qitokens
+    ) Helpers(_connectors) {
+        _addQitokenMapping(_qitokenNames, _tokens, _qitokens);
+    }
```
Comment (thrilok209) [inline contracts/avalanche/mapping/benqi.sol]:
```suggestion
    constructor(
        string[] memory _qitokenNames,
        address[] memory _tokens,
        address[] memory _qitokens
    ) {
        _addQitokenMapping(_qitokenNames, _tokens, _qitokens);
    }
```

File: contracts/avalanche/mapping/benqi.sol [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    mapping (string => TokenMap) public qiTokenMapping;
+
+    modifier isChief {
+        require(msg.sender == instaIndex.master() || connectors.chief(msg.sender), "not-an-chief");
+        _;
+    }
+
+    constructor(address _connectors) {
+        connectors = ConnectorsInterface(_connectors);
+    }
```
Comment (thrilok209) [inline contracts/avalanche/mapping/benqi.sol]:
```suggestion
```

File: contracts/avalanche/static/basic.sol [hunk 11]
```diff
@@ -0,0 +1,77 @@
+pragma solidity ^0.7.0;
```
Comment (thrilok209) [inline contracts/avalanche/static/basic.sol]: We don't need this connector. We can remove it :)

File: contracts/avalanche/mapping/benqi.sol [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+    function getMapping(string memory _tokenId) external view returns (address, address) {
+        TokenMap memory _data = qiTokenMapping[_tokenId];
+        return (_data.token, _data.qitoken);
+    }
+
+}
+
+contract InstaBenqiMapping is Helpers {
```
Comment (thrilok209) [inline contracts/avalanche/mapping/benqi.sol]:
```suggestion
contract InstaBenqiMappingAvalanche is Helpers {
```

File: contracts/avalanche/connectors/benqi/main.sol [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+            tokenInReturn,
+            qiTokenColl,
+            amt,
+            getId,
+            setId
+        );
+    }
+}
+
+contract ConnectV2Benqi is BenqiResolver {
```
Comment (thrilok209) [inline contracts/avalanche/connectors/benqi/main.sol]:
```suggestion
contract ConnectV2BenqiAvalanche is BenqiResolver {
```

File: contracts/avalanche/connectors/Qi/helpers.sol [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+    /**
+     * @dev Reward Token
+     */
+    QiInterface internal constant benqiToken = QiInterface(0x8729438EB15e2C8B576fCc6AeCdA6A148776C0F5);
+
+    /**
+     * @dev Benqi Mapping
+     */
+    BenqiMappingInterface internal constant qiMapping = BenqiMappingInterface(0xe7a85d0adDB972A4f0A4e57B698B37f171519e88);
```
Comment (thrilok209) [inline contracts/avalanche/connectors/Qi/helpers.sol]: TODO: Need to change this after the deployment of `InstaBenqiMappingAvalanche` contract

**Review by thrilok209 — APPROVED** [review]

File: contracts/avalanche/connectors/Qi/helpers.sol (hunk 14 above)
Comment (babkendev01) [reply contracts/avalanche/connectors/Qi/helpers.sol]: @thrilok209

Thank you for your feedback.
Fixed and updated above suggestions except `InstaBenqiMappingAvalanche` contract address change.
Please check.

Comment (thrilok209) [reply contracts/avalanche/connectors/Qi/helpers.sol]: @babkendev01 Awesome! Looks good! Would you be able to deploy `InstaBenqiMappingAvalanche` contract? With all the markets? After that, we can update `InstaBenqiMappingAvalanche` contract address

File: contracts/avalanche/connectors/Qi/main.sol [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        require(benqiToken.delegates(address(this)) != delegatee, "Already delegated to same delegatee.");
+
+        benqiToken.delegate(delegatee);
+
+        _eventName = "LogDelegate(address)";
+        _eventParam = abi.encode(delegatee);
+    }
+}
+
+contract ConnectV2Benqi is BenqiResolver {
```
Comment (thrilok209) [inline contracts/avalanche/connectors/Qi/main.sol]:
```suggestion
contract ConnectV2QiAvalanche is BenqiResolver {
```

File: contracts/avalanche/connectors/benqi/helpers.sol [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+abstract contract Helpers is DSMath, Basic {
+    /**
+     * @dev Benqi Comptroller
+     */
+    ComptrollerInterface internal constant troller = ComptrollerInterface(0x486Af39519B4Dc9a7fCcd318217352830E8AD9b4);
+
+    /**
+     * @dev Benqi Mapping
+     */
+    BenqiMappingInterface internal constant qiMapping = BenqiMappingInterface(0xe7a85d0adDB972A4f0A4e57B698B37f171519e88);
```
Comment (Vaibhav20201) [inline contracts/avalanche/connectors/benqi/helpers.sol]: We have to update the address of InstaBenqiMappingAvalanche here aswell right?

Comment (babkendev01) [reply contracts/avalanche/connectors/benqi/helpers.sol]: Yes, sure. @Vaibhav20201 
We will update it after deploying the mapping contract and will let you know.

**Comment by thrilok209** [comment]
Hey @babkendev01, any timeline on deploying the contracts?

**Comment by babkendev01** [comment]
@thrilok209 @Vaibhav20201 
Deployment done with below setting and mapping address is updated on previous commit. please kindly check for the token names, tokens, addresses and let us know.

_qitokenNames | ["AVAX-A", "BTC-A", "ETH-A", "USDT-A", "LINK-A", "DAI-A", "USDC-A", "QI-A"]
-- | --
_tokens | ["0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE","0x50b7545627a5162F82A992c33b87aDc75187B218","0x49D5c2BdFfac6CE2BFdB6640F4F80f226bc10bAB","0xc7198437980c041c805A1EDcbA50c1Ce5db95118","0x5947BB275c521040051D82396192181b413227A3","0xd586E7F844cEa2F87f50152665BCbc2C279D8d70","0xA7D7079b0FEaD91F3e65f86E8915Cb59c1a4C664","0x8729438EB15e2C8B576fCc6AeCdA6A148776C0F5"]
_qitokens | ["0x5C0401e81Bc07Ca70fAD469b451682c0d747Ef1c","0xe194c4c5aC32a3C9ffDb358d9Bfd523a0B6d1568","0x334AD834Cd4481BB02d09615E7c11a00579A7909","0xc9e5999b8e75C3fEB117F6f73E664b9f3C8ca65C","0x4e9f683A27a6BdAD3FC2764003759277e93696e6","0x835866d37AFB8CB8F8334dCCdaf66cf01832Ff5D","0xBEb5d47A3f720Ec0a390d04b4d41ED7d9688bC7F","0x35Bd6aedA81a7E5FC7A7832490e71F757b0cD9Ce"]

Thank you.

**Comment by thrilok209** [comment]
@babkendev01 I think it's better to have WETH-A and WBTC-A? Instead of just ETH-A and BTC-A?

**Comment by babkendev01** [comment]
@thrilok209 
could you tell me if there is any specific rule for that naming? if it should be the same as token name on Avalanche C Chain, correct name is WBTC.e, WETH.e. please let us know the correct name for the deployment again. thank you.

**Comment by thrilok209** [comment]
> @thrilok209 could you tell me if there is any specific rule for that naming? if it should be the same as token name on Avalanche C Chain, correct name is WBTC.e, WETH.e. please let us know the correct name for the deployment again. thank you.

@babkendev01 TokenId for wrapped token generally has "W" in the start. So, I was suggestion to keep them as "WBTC-A" and "WETH-A"

**Comment by babkendev01** [comment]
> > @thrilok209 could you tell me if there is any specific rule for that naming? if it should be the same as token name on Avalanche C Chain, correct name is WBTC.e, WETH.e. please let us know the correct name for the deployment again. thank you.
> 
> @babkendev01 TokenId for wrapped token generally has "W" in the start. So, I was suggestion to keep them as "WBTC-A" and "WETH-A"

@thrilok209 here is the change. please check.

**Comment by thrilok209** [comment]
> > > @thrilok209 could you tell me if there is any specific rule for that naming? if it should be the same as token name on Avalanche C Chain, correct name is WBTC.e, WETH.e. please let us know the correct name for the deployment again. thank you.
> > 
> > 
> > @babkendev01 TokenId for wrapped token generally has "W" in the start. So, I was suggestion to keep them as "WBTC-A" and "WETH-A"
> 
> @thrilok209 here is the change. please check.

@babkendev01 Awesome! Looks good! We can deploy the connectors next :)

**Comment by babkendev01** [comment]
@thrilok209 
connectors will be deployed on your side or it needs to be done on our side?

**Comment by thrilok209** [comment]
> @thrilok209 connectors will be deployed on your side or it needs to be done on our side?

@babkendev01 connectors will be deployed from your side.

**Comment by babkendev01** [comment]
@thrilok209 
BENQI-A : 0xff5A7A16619d3F4c4Bdc2fEF7AF8A9d0b6c35Ca8
QI-A: 0xaC33E98174A897ECF21BD5C6E67696007AA58527
please check and let us know.

**Comment by thrilok209** [comment]
@babkendev01 Perfect! Connectors look good! We will soon whitelist it. Thank you!

**Comment by thrilok209** [comment]
Connector whitelisted: https://snowtrace.io/tx/0x2ae5aff21fb6bda98e1aee0630a0f0d5904ac27cde22a3cfb8791b74e66f216e

**Pull request merged**